apps/web/pages/api/teams/create.ts

```diff
@@ -0,0 +1,90 @@
+import type { NextApiRequest, NextApiResponse } from "next";
+import type Stripe from "stripe";
+import { z } from "zod";
+
+import stripe from "@calcom/features/ee/payments/server/stripe";
+import { HttpError } from "@calcom/lib/http-error";
+import { defaultHandler, defaultResponder } from "@calcom/lib/server";
+import prisma from "@calcom/prisma";
+import { MembershipRole } from "@calcom/prisma/enums";
+
+const querySchema = z.object({
+  session_id: z.string().min(1),
+});
+
+const checkoutSessionMetadataSchema = z.object({
+  teamName: z.string(),
+  teamSlug: z.string(),
+  userId: z.string().transform(Number),
+});
+
+const generateRandomString = () => {
+  return Math.random().toString(36).substring(2, 10);
+};
+
+async function handler(req: NextApiRequest, res: NextApiResponse) {
+  const { session_id } = querySchema.parse(req.query);
+
+  const checkoutSession = await stripe.checkout.sessions.retrieve(session_id, {
+    expand: ["subscription"],
+  });
+  if (!checkoutSession) throw new HttpError({ statusCode: 404, message: "Checkout session not found" });
+
+  const subscription = checkoutSession.subscription as Stripe.Subscription;
+
+  if (checkoutSession.payment_status !== "paid")
+    throw new HttpError({ statusCode: 402, message: "Payment required" });
+
+  // Let's query to ensure that the team metadata carried over from the checkout session.
+  const parseCheckoutSessionMetadata = checkoutSessionMetadataSchema.safeParse(checkoutSession.metadata);
+
+  if (!parseCheckoutSessionMetadata.success) {
+    console.error(
+      "Team metadata not found in checkout session",
+      parseCheckoutSessionMetadata.error,
+      checkoutSession.id
+    );
+
+    if (!checkoutSession.metadata.userId) {
+      throw new HttpError({
+        statusCode: 400,
+        message: "Can't publish team/org without userId",
+      });
+    }
+  }
+
+  const checkoutSessionMetadata = parseCheckoutSessionMetadata.data ?? {
+    teamName: checkoutSession?.metadata?.teamName ?? generateRandomString(),
+    teamSlug: checkoutSession?.metadata?.teamSlug ?? generateRandomString(),
+    userId: checkoutSession.metadata.userId,
+  };
```

Comment:
Very unlikely that the metadata from creating the checkout session won't be carried over but I wanted to think about this edge case. Do we think this is worth covering?

Reply:
I'd rather let it fail than generating a random string.

Reply:
I was thinking about this as well. At this point the team is already paid for. The user is redirected back to the profile page, where they can see if they need to change the team name.

If we fail, then we would have to make another call to Stripe to cancel the subscription. I think that would be more friction than making the user change the team name.